### PR TITLE
Lf 2804 read only view

### PR DESCRIPTION
**Pull request by wasswarichard**
https://lite-farm.atlassian.net/browse/LF-2804 Read only view, Irrigation task refactor

### CODE REVIEW

**Comment by anant00729** [comment]
Tested. this worked for me. here are the screenshots:

On creating irrigation task
Screen Shot 2022-12-07 at 4 59 15 PM
Screen Shot 2022-12-07 at 4 59 27 PM

Read Only view of irrigation task
Screen Shot 2022-12-07 at 4 59 51 PM

**Comment by anant00729** [comment]
@Duncan-Brain Please review

**Review by anant00729 — APPROVED** [review]

File: packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    table.string('location_size_unit');
+    table.float('percentage_location_irrigated');
+    table.string('percentage_location_irrigated_unit');
+    table.boolean('default_location_flow_rate').defaultTo(false);
+    table.boolean('default_location_application_depth').defaultTo(false);
+    table.boolean('set_default_irrigation_task_type_location').defaultTo(false);
+    table.boolean('set_default_irrigation_task_type_measurement').defaultTo(false);
+  });
+};
+
```
Comment (JimmyRowland) [inline packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js:14]: Please follow other tasks.

**Review by JimmyRowland — CHANGES_REQUESTED** [review]

File: packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+export const up = async function (knex) {
+  await knex.schema.alterTable('irrigation_task', (table) => {
+    table.float('irrigated_area');
```
Comment (JimmyRowland) [inline packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js:3]: t.decimal('irrigated_area', 36, 12);

File: packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js (hunk 1 above)
Comment (wasswarichard) [reply packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js:14]: This is for setting checkbox value

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx [hunk 3]
```diff
@@ -7,7 +7,7 @@ import Button from '../../Form/Button';
 import PropTypes from 'prop-types';
 import Unit, { getUnitOptionMap } from '../../Form/Unit';
 import {
-  estimated_duration,
+  estimated_duration as estimated_time_duration,
```
Comment (JimmyRowland) [inline packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:10]: irrigation_task_estimated_duration
same with other units db_table_column_name

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx [hunk 4]
```diff
@@ -38,43 +37,31 @@ const WaterUseVolumeCalculator = ({ system, setTotalWaterUsage, totalWaterUsage,
...
-        )
-          return estimated_flow_rate * estimated_irrigation_duration * 60;
+        if (estimated_flow_rate_unit.label === 'l/h' && estimated_duration_unit.label === 'h')
+          return estimated_flow_rate * estimated_duration;
+        if (estimated_flow_rate_unit.label === 'l/m' && estimated_duration_unit.label === 'min')
+          return estimated_flow_rate * estimated_duration;
+        if (estimated_flow_rate_unit.label === 'l/h' && estimated_duration_unit.label === 'min')
+          return (estimated_flow_rate * estimated_duration) / 60;
+        if (estimated_flow_rate_unit.label === 'l/m' && estimated_duration_unit.label === 'h')
+          return estimated_flow_rate * estimated_duration * 60;
```
Comment (JimmyRowland) [inline packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:54]: estimated_flow_rate and estimated_duration should give values in irrigation_task_estimated_duration.databaseUnit and irrigation_task_estimated_flow_rate.databaseUnit.
Something might have gone seriously wrong

**Review by JimmyRowland — CHANGES_REQUESTED** [review]

File: packages/api/src/models/irrigationTaskModel.js [hunk 5]
```diff
@@ -45,6 +45,15 @@ class IrrigationTaskModel extends Model {
         application_depth: { type: 'number' },
         application_depth_unit: { type: 'string' },
         default_measuring_type: { type: 'string' },
+        irrigated_area: { type: 'number' },
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/api/src/models/irrigationTaskModel.js:48]: This is a string not an Id

File: packages/api/src/controllers/taskController.js [hunk 6]
```diff
@@ -600,6 +600,17 @@ const taskController = {
       return res.status(400).send({ error });
     }
   },
+  async getIrrigationTaskTypes(req, res) {
+    const { farm_id } = req.params;
+    try {
+      const irrigationTaskTypes = await IrrigationTypesModel.getAllIrrigationTaskTypesByFarmId(
+        farm_id,
+      );
+      res.status(200).json(irrigationTaskTypes);
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/api/src/controllers/taskController.js:609]: @JimmyRowland  that's the end of the function, It's called by the endpoint to return data

File: packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    table.string('irrigated_area_unit');
+    table.string('location_size_unit');
+    table.float('percentage_location_irrigated');
+    table.string('percentage_location_irrigated_unit');
+    table.boolean('default_location_flow_rate').defaultTo(false);
+    table.boolean('default_location_application_depth').defaultTo(false);
+    table.boolean('set_default_irrigation_task_type_location').defaultTo(false);
+    table.boolean('set_default_irrigation_task_type_measurement').defaultTo(false);
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js]: @JimmyRowland that has been updated

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx (hunk 3 above)
Comment (wasswarichard) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:10]: @JimmyRowland  this is a unit used by other components. I can't rename it I can as only type it.

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx [hunk 8]
```diff
@@ -38,43 +37,31 @@ const WaterUseVolumeCalculator = ({ system, setTotalWaterUsage, totalWaterUsage,
   const {
     estimated_flow_rate,
     estimated_flow_rate_unit,
-    estimated_irrigation_duration,
-    estimated_irrigation_duration_unit,
-  } = getValues();
+    estimated_duration,
+    estimated_duration_unit,
+  } = getValues().irrigation_task;
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:42]: @JimmyRowland this has been resolved.

File: packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+export const up = async function (knex) {
+  await knex.schema.alterTable('irrigation_task', (table) => {
+    table.float('irrigated_area');
+    table.string('irrigated_area_unit');
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js:4]: @JimmyRowland this has been resolved

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx [hunk 10]
```diff
@@ -38,43 +37,31 @@ const WaterUseVolumeCalculator = ({ system, setTotalWaterUsage, totalWaterUsage,
...
+          return estimated_flow_rate * estimated_duration;
+        if (estimated_flow_rate_unit.label === 'l/h' && estimated_duration_unit.label === 'min')
+          return (estimated_flow_rate * estimated_duration) / 60;
+        if (estimated_flow_rate_unit.label === 'l/m' && estimated_duration_unit.label === 'h')
+          return estimated_flow_rate * estimated_duration * 60;
         return totalWaterUsage;
       });
     }
-  }, [estimated_irrigation_duration, estimated_flow_rate]);
+  }, [estimated_duration, estimated_flow_rate]);
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:58]: @JimmyRowland this has been resolved

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx [hunk 11]
```diff
@@ -91,9 +78,9 @@ const WaterUseVolumeCalculator = ({ system, setTotalWaterUsage, totalWaterUsage,
         system={system}
         control={control}
         onChangeUnitOption={(e) => {
-          if (e.value === 'l/h' && estimated_flow_rate_unit.value === 'l/min')
+          if (e.value === 'l/h' && estimated_flow_rate_unit?.value === 'l/min')
             setValue(FLOW_RATE, convert(getValues(FLOW_RATE)).from('l/min').to('l/h'));
-          if (e.value === 'l/min' && estimated_flow_rate_unit.value === 'l/h')
+          if (e.value === 'l/min' && estimated_flow_rate_unit?.value === 'l/h')
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:83]: @JimmyRowland this has been resolved with convert unit library

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx [hunk 12]
```diff
@@ -140,22 +127,23 @@ const WaterUseVolumeCalculator = ({ system, setTotalWaterUsage, totalWaterUsage,
 const WaterUseDepthCalculator = ({ system, setTotalWaterUsage, totalWaterUsage, formState }) => {
   const { t } = useTranslation();
   const { register, getValues, watch, control, setValue } = formState();
-  const { irrigated_area, application_depth, percentage_location_irrigated } = getValues();
+  const { irrigated_area, application_depth, percentage_location_irrigated } =
+    getValues().irrigation_task;
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:131]: @JimmyRowland this has been resolved and implemented with a watch

File: packages/webapp/src/components/Task/PureIrrigationTask/index.jsx [hunk 13]
```diff
@@ -1,147 +1,98 @@
...
-    formState: { isValid, errors },
-  } = useForm({
-    mode: 'onChange',
-    shouldUnregister: false,
-    defaultValues: {
-      measurement_type: '',
-      ...persistedFormData,
-    },
-  });
   const { estimated_water_usage_unit, estimated_water_usage } = getValues();
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Task/PureIrrigationTask/index.jsx:42]: @JimmyRowland this has been resolved

File: packages/webapp/src/components/Task/PureIrrigationTask/index.jsx [hunk 14]
```diff
@@ -1,147 +1,98 @@
...
   const handleModalSubmit = () => {
     setValue(ESTIMATED_WATER_USAGE, totalWaterUsage);
     setValue(ESTIMATED_WATER_USAGE_UNIT, getUnitOptionMap()['l']);
     onDismissWaterUseCalculatorModel();
   };
+  const selectedIrrigationTypeOption = useMemo(() => {
+    return IrrigationTypeOptions.filter(
+      (options) => options.value === getValues(IRRIGATION_TYPE),
+    )[0];
+  }, [getValues(IRRIGATION_TYPE), IrrigationTypeOptions]);
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Task/PureIrrigationTask/index.jsx:90]: @JimmyRowland this has been resolved.

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx (hunk 4 above)
Comment (wasswarichard) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx]: @JimmyRowland this has been resolved

File: packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js (hunk 1 above)
Comment (wasswarichard) [reply packages/api/db/migration/20221207131355_add_extra_columns_irrigation_table.js:14]: @JimmyRowland this has been resolved

File: packages/webapp/src/components/Task/PureIrrigationTask/index.jsx [hunk 15]
```diff
@@ -1,147 +1,98 @@
...
-  });
   const { estimated_water_usage_unit, estimated_water_usage } = getValues();
+  const { irrigationTaskTypes = [] } = useSelector(irrigationTaskTypesSliceSelector);
 
-  const stateController = () => {
-    return { register, getValues, watch, control, setValue };
-  };
+  const dispatch = useDispatch();
+  useEffect(() => {
+    dispatch(getIrrigationTaskTypes());
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Task/PureIrrigationTask/index.jsx]: @JimmyRowland this has been resolved.

File: packages/webapp/src/containers/irrigationTaskTypesSlice.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
+import { createSlice } from '@reduxjs/toolkit';
+
+const IrrigationTaskTypesSlice = createSlice({
+  name: 'irrigationTaskTypesReducer',
+  initialState: {
+    loading: false,
+    irrigationTaskTypes: [],
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/containers/irrigationTaskTypesSlice.js:10]: @JimmyRowland this has been resolved.

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx (hunk 3 above)
Comment (JimmyRowland) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:10]: The convention is to create new ones for each field in each table as allowed units might be different.

File: packages/webapp/src/components/Task/PureIrrigationTask/index.jsx [hunk 17]
```diff
@@ -155,16 +106,18 @@ export default function PureIrrigationTask({
...
         )}
       />
       {(irrigationTypeValue === 'OTHER' ||
         getValues(IRRIGATION_TYPE)?.label === t('ADD_TASK.IRRIGATION_VIEW.TYPE.OTHER')) && (
         <Input
           style={{ marginTop: '6px' }}
+          disabled={disabled}
           label={t('ADD_TASK.IRRIGATION_VIEW.WHAT_TYPE_OF_IRRIGATION')}
-          hookFormRegister={register(CREATE_IRRIGATION_TYPE, {
+          hookFormRegister={register(IRRIGATION_TYPE_OTHER, {
```
[the remark this answers is not in the document]
Comment (wasswarichard) [reply packages/webapp/src/components/Task/PureIrrigationTask/index.jsx:111]: @JimmyRowland I didn't explicitly understand this but I think the change I made is what you wanted.

**Review by JimmyRowland — CHANGES_REQUESTED** [review]

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (JimmyRowland) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:63]: I mean there should never be manual unit conversion. Unit conversions are done internally in Unit component.
Say you have a unit component `estimated_duration = 1`, `estimated_duration_unit = s`, and database unit for time is `day`.
The unit component shows 86400s, 
If you update estimated_duration_unit to min,
Then unit component should show 86400 min.
but estimated_duration will be 60.
The following is the previous version of unit component
https://5fa09c09b8435e0022120d64-nyzquumohg.chromatic.com/?path=/story/components-input-unit--square-meter-area-total-area
The new one should work in similar fashion, or unit component needs to be fixed

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (JimmyRowland) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx]: Unit comonent use convert unit internally to sync display value/ unit with real value

**Review by JimmyRowland — CHANGES_REQUESTED** [review]

File: packages/webapp/src/components/Task/PureIrrigationTask/index.jsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (JimmyRowland) [reply packages/webapp/src/components/Task/PureIrrigationTask/index.jsx:111]: image

It seems the pattern has changed. That dropdown was used to create new types

File: packages/api/src/models/irrigationTaskModel.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (JimmyRowland) [reply packages/api/src/models/irrigationTaskModel.js:48]: There should be another migration to convert type to irrigation_type_id.
First add the following types to `irrigation_type` table
['sprinkler'::text, 'drip'::text, 'subsurface'::text, 'flood'::text, 'HAND_WATERING'::text, 'CHANNEL'::text, 'DRIP'::text, 'FLOOD'::text, 'PIVOT'::text, 'SPRINKLER'::text, 'SUB_SURFACE'::text]
Then add an `irrigation_type_id` column and fill the ids according to `irrigation_task.type` and `irrigation_type.irrigation_type_name` and `irrigation_type.irrigation_type_id`.
Then drop irrigation_task.task

**Review by JimmyRowland — CHANGES_REQUESTED** [review]

File: packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (JimmyRowland) [reply packages/webapp/src/components/Modals/WaterUsageCalculatorModal/index.jsx:63]: under pwa branch, you should be able to `pnpm storybook` to test unit component test cases

**Comment by wasswarichard** [comment]
@JimmyRowland  please look at the latest changes,  you still looking at the old version

**Pull request closed without merging**